core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1213,8 +1238,7 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
 
     val unauthorizedForCreateTopicMetadata = unauthorizedForCreateTopics.map(topic =>
-      metadataResponseTopic(Errors.TOPIC_AUTHORIZATION_FAILED, topic, isInternal(topic), util.Collections.emptyList()))
-
+      metadataResponseTopic(Errors.TOPIC_AUTHORIZATION_FAILED, topic, metadataCache.getTopicId(topic), isInternal(topic), util.Collections.emptyList()))
```

Comment:
I think we only care about authorizedForCreate when using topic names. but just want to make sure we don't give a name/id away

Reply:
@jolshan , Thank you, I misunderstood your comments before, I added the following changes:
1. If we are using topicId in metadataReq, we set topicName=null and error=UNKNOWN_TOPIC_ID on unauthorized error
2. If we are using topicName in metadataReq, we set topicId=ZERO and error=TOPIC_AUTHORIZATION_FAILED on unauthorized error.
3. Add a unit test for the above 2 cases.
@rajinisivaram @jolshan , pleas take a look.

Reply:
Quick question. Do we also set topic null when we have actual unknown topic IDs?